web/static/js/components/simulation/VoiceWindow.jsx

```diff
@@ -122,9 +139,121 @@ const VoiceWindow = translate()(class extends Component<VoiceWindowProps, VoiceW
         </div>
       </div>
 
-      <audio ref={audio => { if (audio) this.audio = audio }} preload />
+      <audio ref={audio => { this.setAudio(audio) }} preload />
     </div>
   }
+
+  setAudio(element) {
+    if (!element) return
+    this.audio = element
+  }
 })
 
+class VoiceSpectrum {
+  audioContext: AudioContext
+  audioData: AnalyserNode
+  amplitudeData: Uint8Array
+  canvas: HTMLCanvasElement
+  context: CanvasRenderingContext2D
+  sizes: { width: number, height: number, halfHeight: number }
+  playing: boolean
+
+  constructor() {
+    this.audioContext = new AudioContext()
+    // $FlowFixMe: shut up, AnalyserNode's constructor does take arguments: https://developer.mozilla.org/en-US/docs/Web/API/AnalyserNode/AnalyserNode
+    this.audioData = new AnalyserNode(this.audioContext, { fftSize: 256 })
+    this.amplitudeData = new Uint8Array(0)
+    this.playing = false
+  }
+
+  setCanvas(canvas) {
+    if (canvas) {
+      this.canvas = canvas
+      this.context = canvas.getContext('2d')
+
+      // fix canvas' size for HiDPI from its CSS size (DPI=1)
+      const ratio = window.devicePixelRatio || 1
+      const width = canvas.offsetWidth
+      const height = canvas.offsetHeight
+      canvas.width = canvas.offsetWidth * ratio
+      canvas.height = canvas.offsetHeight * ratio
+      this.context.scale(ratio, ratio)
+
+      this.sizes = {
+        width,
+        height,
+        halfHeight: height / 2
+      }
+    }
+  }
+
+  start(audio) {
+    const audioSource = this.audioContext.createMediaElementSource(audio)
+    audioSource.connect(this.audioContext.destination)
+    audioSource.connect(this.audioData)
```

Comment:
From an example I was reading (https://github.com/mdn/voice-change-o-matic/blob/gh-pages/scripts/app.js#L128-L205) I thought we should do something like this:

```
audioSource.connect(this.audioData)
this.audioData.connect(this.audioContext.destination)
```
Maybe is the same? (or at least, it has the same result?) 🤔

Reply:
Oh, you got me wondering why it worked: AudioAnalyser doesn't process the sound (its output is equivalent to its input) and merely provides an interface to extract data about the audio. Connecting the audio source to both the analyser and the destination, or connecting through the analyser are equivalent :nerd_face:

Yet, it would be different with a Gain or Equalizer that would process/transform the audio, so it's better to properly do the connection!

Reply:
yeah it seemed that it was equivalent. Great!